### PR TITLE
Validate minimal reference expressions

### PR DESCRIPTION
**Pull request by apmoriarty**
Added visitor to validate minimal reference expressions within a query tree.

For use when https://github.com/NationalSecurityAgency/datawave/pull/1274 is merged.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        isValid = false;
+        log.info("Invalid node (" + cause + "): " + JexlStringBuildingVisitor.buildQueryWithoutParse(node));
+    }
+    
+    /*
+     * Pass through
+     */
+    
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: These are the same implementation as the parent class. These can just be removed I think.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        return data;
+    }
+    
+    /*
+     * Short circuits
+     */
+    
+    @Override
+    public Object visit(ASTBlock node, Object data) {
+        return data;
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java:232]: It looks like we could let these delegate to the parent implementations as well. Though here we're short circuiting I guess. Is that the intent? Are these parts of JEXL we know we don't utilize? 

Not saying we do it here, but if there is a subset of the language we're not supporting, we might bubble these "no-ops" up into our BaseVisitor.

Comment (jschmidt10) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java:232]: Oh, just saw the comment above. Missed that initially.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (isValid && isAnyChildAWrappedSingleTerm(node)) {
+            // found a single term child that is wrapped
+            invalidate(node, "wrapped single child");
+            return data;
+        }
+        node.childrenAccept(this, data);
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Do you want to short circuit here if `isValid` is false?

Comment (jschmidt10) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Actually, I like your explicit short-circuit in the And method below. I'd recommend you use that same pattern here. It makes it clear that we are short-circuiting. I'd even say do it the same way in your Reference method above. That way it's the same across all of them.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    /*
+     * Operations
+     */
+    
+    @Override
+    public Object visit(ASTReference node, Object data) {
+        
+        if (isValid && isParen(node) && isParen(grandChild(node))) {
+            // found a double paren
+            invalidate(node, "double paren");
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: The invalidate messages probably deserve an Enum or minimally to be constants.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+     * Helper methods
+     */
+    
+    // is this node a parentheses
+    private boolean isParen(JexlNode node) {
+        return node instanceof ASTReference && node.jjtGetNumChildren() == 1 && node.jjtGetChild(0) instanceof ASTReferenceExpression;
+    }
+    
+    // fetch the grandchild
+    private JexlNode grandChild(JexlNode node) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Might be worth clarifying `getFirstGrandchild` or `getLeftmostGrandchild` or something? Just to specify that you are targeting a specific one out of possibly many.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Works for me. I'll leave the comment in for which nodes are explicitly pass-through

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Good call, I'll clean this up

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java (hunk 4 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Really like this idea!

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java (hunk 5 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Worth doing, thanks for pointing this out!

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Actually, I'll remove the visit for ASTJexlScript and modify the others to only continue recursing through the tree if it's still valid. Opportunity to short circuit on reference expressions, simple nodes, and not nodes

Comment (jschmidt10) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/MinimalReferenceExpressionsVisitor.java]: Cool. Sounds good.

**Review by FineAndDandy — COMMENTED** [review]
I'm ready to approve once @jschmidt10's comments are addressed

**Review by jschmidt10 — APPROVED** [review]

**Review by FineAndDandy — APPROVED** [review]

**Pull request merged**